editor/osm_auth.cpp

```diff
@@ -0,0 +1,418 @@
+  HTTPClientPlatformWrapper request(m_baseUrl + "/logout");
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  return AuthResult::OK;
+}
+
+// Signs a user id using login and password.
+OsmOAuth::AuthResult OsmOAuth::LoginUserPassword(string const & login, string const & password, SessionID const & sid) const
+{
+  map<string, string> params;
+  params["username"] = login;
+  params["password"] = password;
+  params["referer"] = "/";
+  params["commit"] = "Login";
+  params["authenticity_token"] = sid.m_token;
+  HTTPClientPlatformWrapper request(m_baseUrl + "/login");
+  request.set_body_data(buildPostRequest(params), "application/x-www-form-urlencoded");
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  if (!request.was_redirected())
+    return AuthResult::FailLogin;
+  // Since we don't know whether the request was redirected or not, we need to check page contents.
+  return isLoggedIn(request.server_response()) ? AuthResult::OK : AuthResult::FailLogin;
+}
+
+// Signs a user in using a facebook token.
+OsmOAuth::AuthResult OsmOAuth::LoginSocial(string const & callbackPart, string const & socialToken, SessionID const & sid) const
+{
+  string const url = m_baseUrl + callbackPart + socialToken;
+  HTTPClientPlatformWrapper request(url);
+  request.set_cookies(sid.m_cookies);
+  if (!request.RunHTTPRequest())
+    return AuthResult::NetworkError;
+  if (request.error_code() != 200)
+    return AuthResult::ServerError;
+  if (!request.was_redirected())
+    return AuthResult::FailLogin;
+  return isLoggedIn(request.server_response()) ? AuthResult::OK : AuthResult::FailLogin;
+}
+
+// Fakes a buttons press, so a user accepts requested permissions.
+string OsmOAuth::SendAuthRequest(string const & requestTokenKey, SessionID const & sid) const
+{
+  map<string, string> params;
+  params["oauth_token"] = requestTokenKey;
+  params["oauth_callback"] = "";
+  params["authenticity_token"] = sid.m_token;
+  params["allow_read_prefs"] = "yes";
+  params["allow_write_api"] = "yes";
+  params["allow_write_gpx"] = "yes";
+  params["allow_write_notes"] = "yes";
+  params["commit"] = "Save changes";
+  HTTPClientPlatformWrapper request(m_baseUrl + "/oauth/authorize");
+  request.set_body_data(buildPostRequest(params), "application/x-www-form-urlencoded");
+  request.set_cookies(sid.m_cookies);
+
+  if (!request.RunHTTPRequest())
+    return string();
+
+  string const callbackURL = request.url_received();
+  string const vKey = "oauth_verifier=";
+  auto const pos = callbackURL.find(vKey);
+  if (pos == string::npos)
+    return string();
+  auto const end = callbackURL.find("&", pos);
+  return callbackURL.substr(pos + vKey.length(), end == string::npos ? end : end - pos - vKey.length()+ 1);
```

Comment:
@Zverik It's a bug, probably your google auth didn't work because of it.
@ygorshenin good catch, thanks!